### PR TITLE
[scripts] add some types (unfinished)

### PR DESCRIPTION
**Pull request by danpovey**

### CODE REVIEW

**Comment by danpovey** [comment]
This is my proposal for the types to use for data.
@YiwenShaoStephen, can you please look over this ASAP?
This code is mostly just comment without implementation, but let me know if you think it's workable and if you think there is a better way to do this.

File: scripts/waldo/data_types.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+""" TODO
+"""
+
+
+def validate_image_with_mask(x, num_classes = -1):
+    """This function validates an object x that is supposed to represent an image
+    with the corresponding object mask.  It returns None if successful;
+    otherwise it raises an exception.  Specifically it is checking that:
+      x['img'] is a numpy array
```
Comment (danpovey) [inline scripts/waldo/data_types.py]: note: we probably need some more details here, e.g. about what the shape of the numpy array means, and about what happens for color images.  this is too un-detailed as it is.  But let's decide on the big picture first; I'm waiting for input for Yiwen and Hossein and hope to merge to-morrow.

**Comment by YiwenShaoStephen** [comment]
Are these functions for data preparation purpose?

**Comment by aarora8** [comment]
I guess this task is similar to mask generation task done for madcat arabic images. There to fill all points inside a polygon (rectangle in that case), it looped over all points in the bounding box image. Since the madcat images are high resolution images (400k points in a bounding box), it resulted in making the code slow, as python is very good with loops. For 42k madcat images after speedup and parallelization, it can take around 18 hrs. 
Thought it might not be necessary in this case, as for madcat case we had the overlapping bounding box constraint. But if possible should we do this task in C++.

**Comment by YiwenShaoStephen** [comment]
Ok, I get it. BTW, to provide an alternative, see https://www.kaggle.com/c/data-science-bowl-2018#evaluation. They use another method for encoding the mask instead of using polygon idea here. Not sure which one is more feasible and efficient.

**Comment by danpovey** [comment]
The data types are initially for  data preparation but would likely be reusable for the nnet output.

**Comment by danpovey** [comment]
I assume this conversation right now is about enumerating all pixels inside
a polygon.

Regarding efficiency: for the mask generation, at some point it would be
necessary to enumerate all pixels, because we need to create the mask
array.  And I want this to cover non-convex polygons such as bent text, so
the kaggle approach isn't quite general enough.  Let's just get something
working for now and worry more about efficiency later; we can use the
simple approach for regression testing.

I suggest to make the code find all pixels for now, and we can try more
efficient versions later.  It might be possible to use some kind of trick
to do this fast.. for example draw each line in the polygon in such a way
that for each height (i.e. each y value) each line has exactly one x value
present, and then store those x values as lists indexed by y value; and
then use an even/odd approach to fill in the locations between alternating
x values.  Be careful about corners where 2 lines are present.

**Comment by danpovey** [comment]
Actually, let's not rule out dumping the masks to disk somehow.  I don't
know enough about how I/O is done in PyTorch to know exactly how we should
do it, but I'm sure Yiwen will have something to say about that.  (And
let's try to make it possible to use char's or short's to do this, if
Python lets us, as it would take less disk space).

As long as we have the functionality to turn the polygon-based
representation into a mask-based representation, we can always do that in
data preparation instead of during training.

**Comment by danpovey** [comment]
To clarify: when I talk about 'dumping the masks to disk' I am talking about dumping a numpy array of int that contains the object id, like the 'mask' member in Yiwen's nucleus-detection example.  I am assuming that must be standard in object detection.  We could use 'char' for compression if the libraries allow it.

**Comment by YiwenShaoStephen** [comment]
Sure. I will look into PyTorch I/O mechanism later to see if it is possible.

**Comment by aarora8** [comment]
Ok, thank you. Got it.

________________________________

**Comment by danpovey** [comment]
Regarding pytorch I/O:
it seems to like to dump the images and labels together in a tar file, from
looking at you process_data.py in egs/dsb2018/v1/local/.  One possibility
for MADCAT is to have a data preprocessing stage, similar to your
process_data.py but run in parallel, in which we would downsample the data
slightly (e.g. by a factor of 2 or 4) and at the same time compute the
masks, and dump in a tar file like PyTorch likes to do.  Ashish, assume for
the time being that that's what we'll do.

**Comment by aarora8** [comment]
Ok, thanks, will add process_data.py and an option for downsampling images. For polygon mask generation, can we have an overlapping polygon situation.

**Comment by danpovey** [comment]
Yes there can be overlapping polygons, but we'll deal with it by ordering
the polygons from "bottom-most" to "top-most".  I was leaving that till
later.  For now just ignore the issue.  We'll be writing the mask in an
array of the form

 mask[x,y] = object_id

and the top-most polygons will get written last and override previous ones.

**Comment by aarora8** [comment]
ok, thanks, got it.

**Comment by danpovey** [comment]
merging this so it doesn't block anyone.  @aarora8, please finish any TODOs if you get time.

**Comment by aarora8** [comment]
ok, thanks, will do.

**Pull request merged**